### PR TITLE
Scripts and pipelines for basic orchestration

### PR DESCRIPTION
**Pull request by GMNGeoffrey**
This adds the Buildkite pipelines that will orchestrate everything
we run on pre- and post-submit. So far, it just triggers a placeholder
pipeline that sleeps for 10 seconds, but has everything in place to
trigger real pipelines.

With a successful CLA check, the pipeline looks like:
https://buildkite.com/iree/presubmit/builds/49
yes_cla

With a failing CLA check it looks like:
https://buildkite.com/iree/presubmit/builds/34
no_cla
block_prompt

### CODE REVIEW

**Comment by GMNGeoffrey** [comment]
Ok, this still needs a README and it's obviously only a stub (invoking my test pipeline that just sleeps for 10 seconds), but I think it's at a reasonable spot for initial review

File: build_tools/buildkite/wait_for_pipeline_success.py [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+      create_new_build = True
+  else:
+    print("Didn't find previous build for pipeline. Creating a new one.")
+
+  if create_new_build:
+    build = bk.create_build()
+  build_number = get_build_number(build)
+  state = bk.wait_for_build(build_number)
+
+  build_number = get_build_number(build)
```
Comment (pzread) [inline build_tools/buildkite/wait_for_pipeline_success.py:216]: Do we need to get the build_number twice? (line 213)

Comment (GMNGeoffrey) [reply build_tools/buildkite/wait_for_pipeline_success.py:216]: We surely do not :-) Thanks for catching.

File: build_tools/buildkite/wait_for_pipeline_success.py [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+
+
+def main(args):
+  bk = BuildkitePipelineManager.from_environ(args.pipeline)
+  build = bk.get_latest_build()
+  create_new_build = True
+
+  if build:
+    create_new_build = False
+    state = buildkite.BuildState(build["state"])
```
[the remark this answers is not in the document]
Comment (GMNGeoffrey) [reply build_tools/buildkite/wait_for_pipeline_success.py:187]: We do. That waiting should happen with the call to wait_for_build on [line 213](https://github.com/google/iree/pull/8555/files/4cfd4ce16d7e7009677491c8d0dca76872d59e3d#diff-9aedcd988770da9dbbe0cff0d4535d8b729be94308fd93a68292c6530a094a80R214). These conditions here just look at whether to create a new build instead

File: build_tools/buildkite/presubmit.yml [hunk 3]
```diff
@@ -0,0 +1,43 @@
+# Copyright 2022 The IREE Authors
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+# Note: this runs on "security: submitted" agents, since this pipeline itself is
+# submitted code and it fetches the scripts from the main branch.
```
Comment (ScottTodd) [inline build_tools/buildkite/presubmit.yml:8]: So what happens if someone sends a pull request modifying this file?

File: build_tools/buildkite/postsubmit.yml [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+steps:
+  - label: "Execute gcmn-test-pipeline"
+    agents:
+      queue: "orchestration"
+      security: "submitted"
+    commands: |
+      export BUILDKITE_ACCESS_TOKEN="$(gcloud secrets versions access latest --secret=iree-buildkite-presubmit-pipelines)"
+      ./build_tools/buildkite/wait_for_pipeline_success.py gcmn-test-pipeline
```
Comment (ScottTodd) [inline build_tools/buildkite/postsubmit.yml:14]: Is Buildkite automatically setting `BUILDKITE_BRANCH` and the other env vars this script uses? Could add a comment explaining how each option is passed through. (either in the scripts or maybe in a README to explain how the scripts all work together)

File: build_tools/buildkite/presubmit.yml [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    plugins:
+      - thedyrt/skip-checkout#0870506e0b5d4becc164c6e64dbc331938f0bdcf:
+          cd: /tmp/
+    commands: |
+      # TODO: Switch to main branch
+      curl \
+        https://raw.githubusercontent.com/GMNGeoffrey/iree/whole-new-world/build_tools/buildkite/check_cla.py \
+        | python3 - ${BUILDKITE_COMMIT}
+
+  - wait
```
Comment (ScottTodd) [inline build_tools/buildkite/presubmit.yml:29]: Can we bypass the CLA check to run presubmits if someone trusted adds a label / triggers a pipeline manually? Some corporate and individual CLAs can take a while to process.

**Review by ScottTodd — APPROVED** [review]
As the [`build_tools/buildkite`](https://github.com/google/iree/tree/main/build_tools/buildkite) directory gets more files, let's keep an eye on folder structure. Maybe

```
build_tools/
└── buildkite/
    ├── scripts/
    └── pipelines/
        └── cmake/
```

**Comment by GMNGeoffrey** [comment]
> As the [`build_tools/buildkite`](https://github.com/google/iree/tree/main/build_tools/buildkite) directory gets more files, let's keep an eye on folder structure. Maybe
> 
> ```
> build_tools/
> └── buildkite/
>     ├── scripts/
>     └── pipelines/
>         └── cmake/
> ```

Yeah, I decided not to try to do that too soon and see how things shake out first :-)

File: build_tools/buildkite/postsubmit.yml (hunk 4 above)
Comment (GMNGeoffrey) [reply build_tools/buildkite/postsubmit.yml:14]: It is. Documentation incoming

**Review by pzread — APPROVED** [review]

File: build_tools/buildkite/wait_for_pipeline_success.py [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+      create_new_build = True
+    elif args.rebuild == "bad" and state in (
+        buildkite.BuildState.FAILED,
+        buildkite.BuildState.CANCELED,
+        buildkite.BuildState.SKIPPED,
+        buildkite.BuildState.NOT_RUN,
+    ):
+      print(f"Previous build completed with state '{state.name}' and received"
+            f" `--rebuild=bad`, so creating a new one.")
+      create_new_build = True
```
[the remark this answers is not in the document]
Comment (GMNGeoffrey) [reply build_tools/buildkite/wait_for_pipeline_success.py]: Done

File: build_tools/buildkite/presubmit.yml [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+# Note: this runs on "security: submitted" agents, since this pipeline itself is
+# submitted code and it fetches the scripts from the main branch.
+
+steps:
+  # TODO: Is there a better emoji here? Google logo (would have to add to
+  # Buildkite)?
+  - label: ":face_with_monocle: :admission_tickets: :raised_hand: Check CLA"
```
Comment (ScottTodd) [inline build_tools/buildkite/presubmit.yml:13]: Yeah, let's find something other than this set of emojis

image

(not blocking this PR, we can change them later)

**Review by ScottTodd — APPROVED** [review]

File: build_tools/buildkite/presubmit.yml (hunk 7 above)
Comment (GMNGeoffrey) [reply build_tools/buildkite/presubmit.yml:13]: Where'd you get that garbage emoji set?

image

Comment (ScottTodd) [reply build_tools/buildkite/presubmit.yml:13]: :( yeah, it looks better on macOS and Android. Windows 10 looks bad here.

https://emojipedia.org/admission-tickets/
https://emojipedia.org/microsoft/windows-10-anniversary-update/admission-tickets/

**Pull request merged**